packages/data-context/src/data/ProjectConfigManager.ts

```diff
@@ -0,0 +1,790 @@
+import assert from 'assert'
+import pDefer from 'p-defer'
+import type { AllModeOptions, FoundBrowser, FullConfig, OpenProjectLaunchOptions, TestingType } from '@packages/types'
+import debugLib from 'debug'
+import { ChildProcess, ForkOptions, fork } from 'child_process'
+import path from 'path'
+import _ from 'lodash'
+import inspector from 'inspector'
+import chokidar from 'chokidar'
+import { validate as validateConfig, validateNoBreakingConfigLaunchpad, validateNoBreakingConfigRoot, validateNoBreakingTestingTypeConfig } from '@packages/config'
+import type { SetupFullConfigOptions } from './ProjectLifecycleManager'
+import { CypressEnv } from './CypressEnv'
+
+const debug = debugLib(`cypress:lifecycle:ProjectConfigManager`)
+
+const CHILD_PROCESS_FILE_PATH = require.resolve('@packages/server/lib/plugins/child/require_async_child')
+
+const UNDEFINED_SERIALIZED = '__cypress_undefined__'
+
+const POTENTIAL_CONFIG_FILES = [
+  'cypress.config.ts',
+  'cypress.config.mjs',
+  'cypress.config.cjs',
+  'cypress.config.js',
+]
+
+type ProjectConfigManagerOptions = {
+  configFile: string | false
+  projectRoot: string
+  nodePath: string | null | undefined
+  modeOptions: Partial<AllModeOptions>
+  isRunMode: boolean
+  handlers: IpcHandler[]
+  hasCypressEnvFile: boolean
+  cypressVersion: string
+  onError: (cypressError: CypressError, title?: string | undefined) => void
+  onWarning: (err: CypressError) => void
+  toLaunchpad: (...args: any[]) => void
+  onInitialConfigLoaded: (initialConfig: Cypress.ConfigOptions) => void
+  onFinalConfigLoaded: (finalConfig: FullConfig) => Promise<void>
+  updateWithPluginValues: (config: FullConfig, modifiedConfig: Partial<Cypress.ConfigOptions>) => FullConfig
+  initializeActiveProject: (options?: OpenProjectLaunchOptions) => Promise<unknown>
+  setupFullConfigWithDefaults: (config: SetupFullConfigOptions) => Promise<FullConfig>
+  machineBrowsers: () => FoundBrowser[] | Promise<FoundBrowser[]>
+}
+
+type ConfigManagerState = 'pending' | 'loadingConfig' | 'loadedConfig' | 'loadingNodeEvents' | 'ready' | 'destroyed' | 'errored'
+
+export class ProjectConfigManager {
+  private _configFilePath: string | undefined
+  private _cachedFullConfig: FullConfig | undefined
+  private _childProcesses = new Set<ChildProcess>()
+  private _eventsIpc?: ProjectConfigIpc
+  private _eventProcess: ChildProcess | undefined
+  private _requireWatchers: Record<string, chokidar.FSWatcher> = {}
+  private _watchers = new Set<chokidar.FSWatcher>()
+  private _registeredEvents: Record<string, Function> = {}
+  private _registeredEventsTarget: TestingType | undefined
+  private _initializedProject: unknown | undefined
+  private _testingType: TestingType | undefined
+  private _state: ConfigManagerState = 'pending'
+  private _loadConfigPromise: Promise<LoadConfigReply> | undefined
+  private _cachedLoadConfig: LoadConfigReply | undefined
+
+  #cypressEnv: CypressEnv
```

Comment:
Just curious, why prefix these names with `#`?

Reply:
This is the new ES syntax for private class member: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

For some reason I thought TS compiled it to just use a symbol, but i guess it uses a WeakSet to keep track of the class instance 🤔 https://www.typescriptlang.org/play#code/MYGwhgzhAEBiD29oG8BQ1oGIC2BTbARrgE4AUAlCuhtMbgC4CuxAdtAIzUC+qqXQA